# Week 1 Mon: Reset your toolkit: numbers, variables, returns, and compounding

Estimated time: 6 hours

## Why this matters
Quants constantly translate market moves into numbers. If returns and compounding are shaky, every later topic like portfolio optimization, forecasting, and risk measurement becomes confusing.

## Study blocks
- Block 1 (45 min): Reset notation: prices, returns, percentages, and variables.
- Block 2 (60 min): Learn simple return, gross return, and compounding intuition.
- Block 3 (45 min): Work through hand-calculated examples and common traps.
- Block 4 (60 min): Implement returns and wealth paths in Python and pandas.
- Block 5 (30 min): Practice, reflect, and do a short interview drill.

## Concept notes
### Price change versus return
If a stock moves from 100 to 105, the price change is 5, but the return is 5 divided by 100, which is 5 percent. Returns let us compare moves across assets with different price levels.

Technical note: For a one-period simple return, use r_t = P_t / P_(t-1) - 1. The gross return is 1 + r_t = P_t / P_(t-1).

Finance example: A 5-point move means something very different for a stock priced at 20 than for a stock priced at 500. Returns normalize that difference.

### Compounding and the wealth path
Once your capital changes, the next day's gain or loss applies to the new amount, not the old amount. That is why growth is multiplicative.

Technical note: If wealth starts at W_0, then after returns r_1, r_2, ..., r_t, wealth is W_t = W_0 * product(1 + r_i).

Finance example: A loss of 50 percent requires a later gain of 100 percent just to get back to the starting point.

### Why arithmetic intuition can mislead
A gain of 10 percent and a loss of 10 percent do not cancel out because the second move happens on a different base.

Technical note: The path 100 -> 110 -> 99 gives returns of +10 percent and -10 percent, but total wealth falls by 1 percent.

Finance example: This is why traders care about drawdowns and recovery periods, not just average daily returns.


## Code Lab 1: Prices to returns and wealth

In [1]:
import pandas as pd

prices = pd.Series([100, 103, 101, 104, 107], name="price")
returns = prices.pct_change().dropna()
gross_returns = 1 + returns
wealth = 1000 * gross_returns.cumprod()

print("Prices:")
print(prices.to_string(index=False))
print("\nReturns:")
print((returns * 100).round(2).to_string(index=False))
print("\nGross returns:")
print(gross_returns.round(4).to_string(index=False))
print("\nWealth path:")
print(wealth.round(2).to_string(index=False))


Prices:
100
103
101
104
107

Returns:
 3.00
-1.94
 2.97
 2.88

Gross returns:
1.0300
0.9806
1.0297
1.0288

Wealth path:
1030.0
1010.0
1040.0
1070.0


## Code Lab 2: Path dependency

In [2]:
import pandas as pd

paths = pd.DataFrame(
    {
        "scenario": ["up then down", "down then up"],
        "r1": [0.10, -0.10],
        "r2": [-0.10, 0.10],
    }
)
paths["final_wealth_from_100"] = 100 * (1 + paths["r1"]) * (1 + paths["r2"])
print(paths)


       scenario   r1   r2  final_wealth_from_100
0  up then down  0.1 -0.1                   99.0
1  down then up -0.1  0.1                   99.0


## Practice recap
- A stock moves from 80 to 92. What is the simple return?
  Suggested answer: Return = 92 / 80 - 1 = 0.15, so 15 percent.
- You start with 2000 and earn returns of 5 percent, 3 percent, and -2 percent. What is final wealth?
  Suggested answer: 2000 * 1.05 * 1.03 * 0.98 = 2119.32.
- Why is a 30 percent loss harder to recover from than it first sounds?
  Suggested answer: Because the required recovery is measured from the reduced base. After falling from 100 to 70, you need 30 / 70 = 42.86 percent to get back to 100.

## Interview drill
- Q: Why do quants work with returns instead of raw prices?
  A: Returns make moves comparable across assets and line up naturally with portfolio aggregation, statistics, and risk analysis.
- Q: What does compounding change in performance analysis?
  A: It makes the order of gains and losses matter because capital evolves multiplicatively over time.

## 6-Hour Completion Roadmap

Use this minimum structure to turn today's notebook into a serious study session:

- Phase 1 (60 min): Read concept notes and rewrite the core idea from memory.
- Phase 2 (60 min): Complete and extend all code labs with at least one variation.
- Phase 3 (60 min): Do formula retrieval, scenario analysis, and error-log updates.
- Phase 4 (60 min): Practice interview responses aloud.
- Phase 5 (60 min): Implement one robustness check.
- Phase 6 (60 min): Write a concise memo with limitations and next steps.

## Formula Rewrite Drill

In [3]:
import pandas as pd

formula_drill = pd.DataFrame(
    [{'formula': 'r_t=\\frac{P_t}{P_{t-1}}-1', 'from_memory': '', 'term_explanations': ''}, {'formula': '\\ell_t=\\ln\\left(\\frac{P_t}{P_{t-1}}\\right)', 'from_memory': '', 'term_explanations': ''}, {'formula': 'W_T=W_0\\prod_{t=1}^{T}(1+r_t)', 'from_memory': '', 'term_explanations': ''}]
)
print(formula_drill)


                                      formula from_memory term_explanations
0                   r_t=\frac{P_t}{P_{t-1}}-1                              
1  \ell_t=\ln\left(\frac{P_t}{P_{t-1}}\right)                              
2               W_T=W_0\prod_{t=1}^{T}(1+r_t)                              


## Real Market Data Lab (Topic-Aligned)

        Use today's topic (Reset your toolkit: numbers, variables, returns, and compounding) with one concrete dataset and one measurable output.

        - Use curriculum/datasets/real_market_prices.csv as reproducible fallback data.
- Build one table and one chart supporting a decision.
- Document one limitation and one robustness check.

In [4]:
TASK = 'Compute log returns and summarize annualized return/risk with one caveat.'
print('Lab task:', TASK)

from pathlib import Path
import pandas as pd

market = pd.read_csv(Path("curriculum/datasets/real_market_prices.csv"), parse_dates=["date"])
prices = market.pivot(index="date", columns="symbol", values="close").dropna()
ret = prices.pct_change().dropna()
ann_ret = ret.mean() * 252
ann_vol = ret.std() * (252 ** 0.5)
print(pd.DataFrame({"ann_return": ann_ret, "ann_vol": ann_vol}).round(4))


Lab task: Compute log returns and summarize annualized return/risk with one caveat.
        ann_return  ann_vol
symbol                     
GLD         0.1856   0.1770
QQQ         0.1719   0.2247
SPY         0.1524   0.1698
TLT        -0.0683   0.1579


## Real-World Takeaway Prompt

Write 5-8 lines for today's topic (Reset your toolkit: numbers, variables, returns, and compounding):

1. Which symbol looked most volatile and why that matters.
2. Which pair looked most diversifying.
3. One realistic portfolio/risk decision you could make from this table.

## Interview Question + Python Solution Drill

Use this format for interview preparation:

1. State the question clearly.
2. Explain your approach in plain language.
3. Write Python code on relevant data.
4. Interpret one risk caveat in words.

**Drill prompt**
- Compute log returns and summarize annualized return/risk with one caveat.

In [5]:
TASK = 'Compute log returns and summarize annualized return/risk with one caveat.'
print('Interview drill task:', TASK)

from pathlib import Path
import pandas as pd

market = pd.read_csv(Path("curriculum/datasets/real_market_prices.csv"), parse_dates=["date"])
prices = market.pivot(index="date", columns="symbol", values="close").dropna()
ret = prices.pct_change().dropna()
ann_ret = ret.mean() * 252
ann_vol = ret.std() * (252 ** 0.5)
print(pd.DataFrame({"ann_return": ann_ret, "ann_vol": ann_vol}).round(4))


Interview drill task: Compute log returns and summarize annualized return/risk with one caveat.
        ann_return  ann_vol
symbol                     
GLD         0.1856   0.1770
QQQ         0.1719   0.2247
SPY         0.1524   0.1698
TLT        -0.0683   0.1579


## Scenario Analysis Drill

In [6]:
import pandas as pd

scenarios = pd.DataFrame(
    [
        {"scenario": "base_case", "assumption": "", "expected_effect": ""},
        {"scenario": "bull_case", "assumption": "", "expected_effect": ""},
        {"scenario": "stress_case", "assumption": "", "expected_effect": ""},
    ]
)
print("Topic:", 'Reset your toolkit: numbers, variables, returns, and compounding')
print(scenarios)


Topic: Reset your toolkit: numbers, variables, returns, and compounding
      scenario assumption expected_effect
0    base_case                           
1    bull_case                           
2  stress_case                           


## Closed-Book Retrieval and Error Log

In [7]:
import pandas as pd

retrieval_scorecard = pd.DataFrame(
    [
        {"prompt": "Explain today's concept in 3 lines", "score_0_to_5": None, "notes": ""},
        {"prompt": "Write one key formula without notes", "score_0_to_5": None, "notes": ""},
        {"prompt": "Give one realistic failure mode", "score_0_to_5": None, "notes": ""},
        {"prompt": "Connect to one trading/risk decision", "score_0_to_5": None, "notes": ""},
    ]
)

error_log = pd.DataFrame(
    [
        {
            "concept": "",
            "mistake": "",
            "correction": "",
            "next_review_date": "",
        }
    ]
)
print("Retrieval scorecard template:")
print(retrieval_scorecard)
print("\nError log template:")
print(error_log)


Retrieval scorecard template:
                                 prompt score_0_to_5 notes
0    Explain today's concept in 3 lines         None      
1   Write one key formula without notes         None      
2       Give one realistic failure mode         None      
3  Connect to one trading/risk decision         None      

Error log template:
  concept mistake correction next_review_date
0                                            


## Final 30-Minute Deliverable

Write a short memo (150-250 words) with this structure:

1. Core idea of Reset your toolkit: numbers, variables, returns, and compounding in plain language.
2. One technical detail or formula and why it matters.
3. One practical quant use case.
4. One limitation or failure mode and how you would detect it.